### ISSUE TITLE
Parse Errors

### ISSUE THREAD

**Issue by flywire** [marlanperumal/pdf_statement_reader#33]
See #30 for pdf and config:
```
Date    Transaction                             Debit   Credit  Balance
01 Jul  2018 OPENING BALANCE                                    $1,384.89 CR
01 Jul  DEBIT INTEREST CHARGED on this account
        to June 30. 2018 is $0.11
02 Jul  Transfer to another Bank NetBank        372.00          $1,012.89 CR
        Rob Ubank Transfer
```
1. Amount starting/ending/contains CR/DB
1. No year in date
1. Skip lines starting/ending/contains Balance/Forward
1. Concatenate wrapped Transaction description
1. Allow currency sign
1. Allow thousands separator

**Comment by flywire** [comment]
Using https://github.com/marlanperumal/pdf_statement_reader/issues/34#issuecomment-860212042

```
(pdf_statement_reader) E:\venv\pdf_statement_reader>psr pdf2csv MockTemplate.pdf
Converted MockTemplate.pdf and saved as MockTemplate.csv
```

```
Date,Transaction Description,Transaction Detail,Debit Amount,Credit Amount,Balance
,"Lorem ipsum dolor sit amet,","consectetur adipiscing elit, sed",-197.60,,802.40
,tristique senectus. Quisque,,-976.14,,192.73
,egestas diam in arcu curs,euismod quis. Eget mi proin sed,-904.80,,-712.07
,libero enim sed. Dignissim enim,,,327.70,-831.74
,sit amet venenatis urna. Ipsum a,,,362.86,-468.88
,arcu cursus vitae congue mauris,,-455.41,,-924.29
,rhoncus aenean,,,169.89,-754.40
,consectetur adipiscing,,,462.19,-292.21
,arcu felis bibendum ut tristique,,,452.75,160.54
,et egestas. Sagittis vitae et leos,,,171.54,332.08
,duis ut diam. Tempor commo,,,237.17,569.25
,ullamcorper,sed arcu non. Facilisis volutpat,,392.18,961.43
,est velil egestas. Vel facilisi,,-786.68,,595.30
```

1. Date dropped with format used
2. Transactions with thousands separator in amounts dropped
3. Where next transaction dropped, next Transaction Description shown in previous Transaction Detail
4. Only first page processed

**Comment by marlanperumal** [comment]
Can you provide the full json config that you used? In the issue that you link, I only see the layout section

**Comment by flywire** [comment]
> (pdf_statement_reader) E:\venv\pdf_statement_reader>psr pdf2csv MockTemplate.pdf

`[default: za.absa.cheque]`

**Comment by marlanperumal** [comment]
Oh I see did you just use the default config?

```
{
    "layout": {
        "default": {
            "area": [280, 27, 763, 576],
            "columns": [83, 264, 344, 425, 485, 570]
        },
        "first": {
            "area": [480, 27, 763, 576],
            "columns": [83, 264, 344, 425, 485, 570]
        }
    },
    "columns": {
        "trans_date": "Date",
        "trans_type": "Transaction Description",
        "trans_detail": "Transaction Detail",
        "debit": "Debit Amount",
        "credit": "Credit Amount",
        "balance": "Balance"
    },
    "order": [
        "trans_date",
        "trans_type",
        "trans_detail",
        "debit",
        "credit",
        "balance"
    ],
    "cleaning": {
        "numeric": ["debit", "credit", "balance"],
        "date": ["trans_date"],
        "date_format": "%d/%m/%Y",
        "trans_detail": "below",
        "dropna": ["balance"]
    }
}
```

You should create a new json config file, updated with the correct values, store it in `pdf_statement_reader/config/au/citi/mock.json` or similar and call it with

```
psr pdf2csv MockTemplate.pdf --config=au.citi.mock
```

**Comment by flywire** [comment]
I envisaged this would be project sample pdf to use with defaults.

Testing pdf(s) are also needed. Maybe this one could be `pdf_statement_reader/config/00/test/mock.json`

**Comment by flywire** [comment]
lol now I finally get it, we have a cross-communication issue.

> For now the only config supported is for Cheque account statements from Absa bank in South Africa.

You are suggesting a new config file is required but I was trying to create a default pdf.

Testing pdf(s) are also needed. Maybe `pdf_statement_reader/config/00/test/mock.json`

**Comment by flywire** [comment]
Read `MockTemplate.pdf` date with `"date_format": "%d/%m/%y"` but saved in csv file as "yyyy/mm/dd"

**Comment by flywire** [comment]
Clean numeric moves trailing `-` to leading `-`. Also needs:
1. strip spaces, thousand separator, currency symbol
1. replace `DB` with leading `-`

**Comment by flywire** [comment]
@marlanperumal Why isn't p2 being parsed?

**Comment by flywire** [comment]
Need to update JSON config format for [Clean numeric](https://github.com/marlanperumal/pdf_statement_reader/blob/master/pdf_statement_reader/parse.py#L40-L53) something like:

```py
def clean_numeric(df, config):
    numeric_cols = [config["columns"][col] for col in config["cleaning"]["numeric"]]

    def format_negatives(s):
        s = str(s)
        df[col] = df[col].str.upper().replace("DR", "-")
        if s.endswith("-"):
            return "-" + s[:-1]
        else:
            return s

    for col in numeric_cols:
        df[col] = df[col].apply(format_negatives)
        df[col] = df[col].str.upper().replace("CR", "-")
        df[col] = df[col].str.replace(",", "")
        df[col] = df[col].str.replace("$", "")
        df[col] = df[col].str.replace(" ", "")
        df[col] = pd.to_numeric(
            df[col],
            errors="coerce"
        )
```

**Comment by flywire** [comment]
Tidy this up with defaults from JSON to parse currency:

```py
def clean_numeric(df, config):
    numeric_cols = [config["columns"][col] for col in config["cleaning"]["numeric"]]

    def format_currency_number(s):
        s = str(s)
        if s.upper().find("DR") >= 0:
            if s.find("-") >= 0:
                s = s.upper().replace("DR", "")
            else:
                s = s.upper().replace("DR", "-")
        if s.endswith("-"):
            s = "-" + s[:-1]
        s = s.upper().replace('$', '').replace('CR', '')
        return s

    for col in numeric_cols:
        df[col] = df[col].apply(format_currency_number)
        df[col] = df[col].str.replace(",", "")
        df[col] = df[col].str.replace(" ", "")
        df[col] = pd.to_numeric(
            df[col],
            errors="coerce"
        )
    print(df)
```

**Comment by flywire** [comment]
What do you think about changing JSON cleaning as follows?
* `"drop": ["description", "(^(?i)BALANCE.*(?i)FORWARD$)"]`

**Comment by marlanperumal** [comment]
> What do you think about changing JSON cleaning as follows?
> 
> * `"drop": ["trans_type", "(^(?i)BALANCE.*(?i)FORWARD$)"]`

To be honest I'd prefer to be explicit on the column names, also because of my previously mentioned aversion to regex. I might be persuaded to go for some additional keys like
- `drop_if_prefix`
- `drop_if_suffix`
- `drop_if_contains`
-

**Comment by marlanperumal** [comment]
> Read `MockTemplate.pdf` date with `"date_format": "%d/%m/%y"` but saved in csv file as "yyyy/mm/dd"

Are you reading the csv file into Excel? (hence the format change). The desired behaviour is actually to always output the date in iso-format `YYYY-MM-DD` no matter what the input format.

**Comment by marlanperumal** [comment]
> lol now I finally get it, we have a cross-communication issue.
> 
> > For now the only config supported is for Cheque account statements from Absa bank in South Africa.
> 
> You are suggesting a new config file is required but I was trying to create a sample pdf to use with default setting.
> 
> Testing pdf(s) are also needed. Maybe `pdf_statement_reader/config/00/test/mock.json`

I'll try to create a config for your mock pdf later this week. It'll actually be a a really good base to use for testing of all the other functionality

**Comment by flywire** [comment]
> config > [country code] > [bank] > [statement type].json

Country code is irrelevant as bank format normally consistent across countries

> explicit column names

json file already uses field coding names, more portable as not specific to actual statements

> The desired behaviour is actually to always output the date in iso-format `YYYY-MM-DD` no matter what the input format.

No, output format specified by software consuming file

> I'll try to create a config for your mock pdf later this week

Only difference from default is `"date_format": "%d/%m/%Y"` -> `"date_format": "%d/%m/%y"`. Do one for [CbaBankStatement.pdf](https://github.com/marlanperumal/pdf_statement_reader/files/6565881/Commonweath.Bank.Statement.pdf).

**Comment by flywire** [comment]
Not bad for first run.

See https://github.com/marlanperumal/pdf_statement_reader/issues/34#issue-906756112 - probably using https://templatelab.com/file/89621/ with [json config](https://github.com/marlanperumal/pdf_statement_reader/issues/34#issuecomment-864997528).

Note: There are 2 invalid balances in file because `.` is used as both decimal and thousands separator so those values are ignored.

Using `logging.debug`:

```
(psr) E:\venv\psr>psr pdf2csv -c cba.saving CbaBankStatement.pdf
   Unnamed: 0    Date                             Transaction   Debit  Credit  Balance
0         NaN  01 Jul                    2018 OPENING BALANCE     NaN     NaN  1384.89
1         NaN  01 Jul  DEBIT INTEREST CHARGED on this account     NaN     NaN      NaN
2         NaN     NaN               to June 30. 2018 is $0.11     NaN     NaN      NaN
3         NaN  02 Jul        Transfer to another Bank NetBank  372.00     NaN  1012.89
4         NaN     NaN                      Rob Ubank Transfer     NaN     NaN      NaN
5         NaN  02 Jul                  Transfer from MCKAY MJ     NaN    43.8      NaN
6         NaN     NaN              Mick Mckay - Neck hackles.     NaN     NaN      NaN
7         NaN  02 Jul     Direct Debit 000115 COLONIAL MUTUAL   25.00     NaN      NaN
8         NaN     NaN                        1200180874627741     NaN     NaN      NaN
9         NaN  02 Jul                          Loan Repayment  280.00     NaN   751.69
10        NaN     NaN                       LN REPAY694259331     NaN     NaN      NaN
11        NaN  03 Jul      PETSTOCK HEATHMONT P HEATHMONT AUS   32.99     NaN   718.70
12        NaN     NaN                             Card xx4521     NaN     NaN      NaN
13        NaN     NaN                  Value Date: 01/07/2018     NaN     NaN      NaN
14        NaN  03 Jul     WOOLWORTHS 3149 EASTLA RINGWOOD AUS  134.26     NaN   584.44
15        NaN     NaN                             Card xx4521     NaN     NaN      NaN
16        NaN     NaN                  Value Date: 01/07/2018     NaN     NaN      NaN
17        NaN  03 Jul             HEATHMONT IGA HEATHMONT AUS   30.00     NaN   554.44
18        NaN     NaN                             Card xx4521     NaN     NaN      NaN
19        NaN     NaN                  Value Date: 30/06/2018     NaN     NaN      NaN
20        NaN  03 Jul   POST HEATHMONT LPOHE HEATHMONT CA AUS   47.57     NaN   506.87
21        NaN     NaN                             Card xx4521     NaN     NaN      NaN
22        NaN     NaN                  Value Date: 29/06/2018     NaN     NaN      NaN
23        NaN  03 Jul      FIVE STAR MUSIC PL RINGWOOD VI AUS   50.00     NaN   456.87
24        NaN     NaN                             Card xx4521     NaN     NaN      NaN
25        NaN     NaN                  Value Date: 28/06/2018     NaN     NaN      NaN
26        NaN     NaN                 BALANCE CARRIED FORWARD     NaN     NaN   456.87
27        NaN     NaN                 BALANCE BROUGHT FORWARD     NaN     NaN   456.87
28        NaN  03 Jul   POST HEATHMONT LPOHE HEATHMONT CA AUS   55.65     NaN   401.22
29        NaN     NaN                             Card xx4521     NaN     NaN      NaN
30        NaN     NaN                  Value Date: 28/06/2018     NaN     NaN      NaN
31        NaN     NaN                         CLOSING BALANCE     NaN     NaN   401.22
Converted CbaBankStatement.pdf and saved as CbaBankStatement.csv
```

**Comment by flywire** [comment]
Still required:
1. Improved cleaning eg:
> * "drop": ["trans_type", "(^(?i)BALANCE.*(?i)FORWARD$)"]

2. Format output dates
3. Edit pdf input file errors

[CbaBankStatement.csv](https://github.com/marlanperumal/pdf_statement_reader/files/6692500/CbaBankStatement.csv)
TLDR - [CbaBankStatementOut.pdf](https://github.com/marlanperumal/pdf_statement_reader/files/6692435/CbaBankStatementOut.pdf)
[cba.saving.json](https://github.com/marlanperumal/pdf_statement_reader/files/6692441/saving.json.txt)

Custom Code:
```py
def clean_date(df, config):
...
    year = df.iloc[0, 1][0:4]
    for col in date_cols:
        df[col] += ' ' + year
        print(df[col])
        df[col] = pd.to_datetime(
            df[col],
            errors="coerce",
            format=date_format + ' %Y'
        )


def unwrap_trans_detail(df, config):
    trans_detail = config["columns"]["trans_detail"]
    trans_date = config["columns"]["trans_date"]
    j = 0
    for i, row in df.iterrows():
        # if pd.isnull(row['trans_date']):
        if pd.isnull(row[0]):
            df.loc[j, trans_detail] += ' ' + df.loc[i, trans_detail]
        else:
            j = i

...
def parse_statement(filename, config):
...
    if "trans_detail" in config["cleaning"]:
        unwrap_trans_detail(statement, config)
```

**Comment by flywire** [comment]
Issue remains for misidentified data after end of transaction data as [previously described](https://github.com/marlanperumal/pdf_statement_reader/issues/30#issue-904762423). Suggest support dropping all records after user described string.

**Comment by marlanperumal** [comment]
Is this still an issue? I've updated the project (including migrating to uv for python and package management) and fixed a couple bugs along the way. 

I'm picking up this project again after a bit of a hiatus. If I don't see any response on this issue within 2 weeks, it will be closed, so that only current issues will be present and can be addressed.

**Comment by marlanperumal** [comment]
Ok - closing this issue then
